### PR TITLE
A more generic interface for dataflow analysis

### PR DESCRIPTION
**Pull request by ecstatic-morse**
#64470 requires a transfer function that is slightly more complex than the typical `gen`/`kill` one. Namely, it must copy state bits between locals when assignments occur (see #62547 for an attempt to make this fit into the existing framework). This PR contains a dataflow interface that allows for arbitrary transfer functions. The trade-off is efficiency: we can no longer coalesce transfer functions for blocks and must visit each statement individually while iterating to fixpoint.

Another issue is that poorly behaved transfer functions can result in an analysis that fails to converge. `gen`/`kill` sets do not have this problem. I believe that, in order to guarantee convergence, flipping a bit from `false` to `true` in the entry set cannot cause an output bit to go from `true` to `false` (negate all preceding booleans when `true` is the bottom value). Perhaps someone with a more formal background can confirm and we can add a section to the docs?

This approach is not maximally generic: it still requires that the lattice used for analysis is the powerset of values of `Analysis::Idx` for the `mir::Body` of interest. This can be done at a later date. Also, this is the bare minimum to get #64470 working. I've not adapted the existing debug framework to work with the new analysis, so there are no `rustc_peek` tests either. I'm planning to do this after #64470 is merged.

Finally, my ultimate plan is to make the existing, `gen`/`kill`-based `BitDenotation` a special case of `generic::Analysis`. Currently they share a ton of code. I should be able to do this without changing any implementers of `BitDenotation`. Something like:

```rust
struct GenKillAnalysis<A: BitDenotation> {
    trans_for_block: IndexVec<BasicBlock, GenKillSet<A::Idx>>,
    analysis: A,
}

impl<A> generic::Analysis for GenKillAnalysis<A> {
    // specializations of `apply_{partial,whole}_block_effect`...
}
```

r? @pnkfelix

### CODE REVIEW

File: src/librustc_mir/dataflow/mod.rs [hunk 1]
```diff
@@ -30,6 +30,7 @@ use self::move_paths::MoveData;
```
Comment (Centril) [inline src/librustc_mir/dataflow/mod.rs]: btw, as drive-by improvement potential, it would be great if this module pointed to some good beginner->less-beginner resources for understanding dataflow.

File: src/librustc_mir/dataflow/generic.rs [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+///     let mut cursor = dataflow::ResultsCursor::new(body, results);
+///
+///     for statement_index in body.block_data[START_BLOCK].statements.iter() {
+///         cursor.seek_after(Location { block: START_BLOCK, statement_index });
+///         let state = cursor.get();
+///         println!("{:?}", state);
+///     }
+/// }
+/// ```
+pub trait Analysis<'tcx>: BottomValue {
```
Comment (oli-obk) [inline src/librustc_mir/dataflow/generic.rs]: Is this scheme expected to replace our existing dataflow architecture or do we need both schemes?

File: src/librustc_mir/dataflow/generic.rs [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+    /// The index type used to access the dataflow state.
+    type Idx: Idx;
+
+    /// A name describing the dataflow analysis being implemented.
+    ///
+    /// The name should be suitable as part of a filename, so avoid whitespace, slashes or periods
+    /// and try to keep it short.
+    fn name() -> &'static str;
```
Comment (oli-obk) [inline src/librustc_mir/dataflow/generic.rs]: could also be an associated constant

File: src/librustc_mir/dataflow/generic.rs [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+
+    /// The size of each bitvector allocated for each block.
+    fn bits_per_block(&self, body: &mir::Body<'tcx>) -> usize;
+
+    /// Mutates the entry set of the `START_BLOCK` to containthe initial state for dataflow
```
Comment (oli-obk) [inline src/librustc_mir/dataflow/generic.rs]:
```suggestion
    /// Mutates the entry set of the `START_BLOCK` to contain the initial state for dataflow
```

File: src/librustc_mir/dataflow/generic.rs [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+    type Idx: Idx;
+
+    /// A name describing the dataflow analysis being implemented.
+    ///
+    /// The name should be suitable as part of a filename, so avoid whitespace, slashes or periods
+    /// and try to keep it short.
+    fn name() -> &'static str;
+
+    /// The size of each bitvector allocated for each block.
+    fn bits_per_block(&self, body: &mir::Body<'tcx>) -> usize;
```
Comment (oli-obk) [inline src/librustc_mir/dataflow/generic.rs]: why does this depend on the body?

File: src/librustc_mir/dataflow/generic.rs [hunk 6]
```diff
@@ -0,0 +1,496 @@
+#![allow(unused)]
```
Comment (oli-obk) [inline src/librustc_mir/dataflow/generic.rs]: you can also make it public, there's a good chance we'll use it from clippy

**Review by oli-obk — COMMENTED** [review]
I like it. This is much clearer (to me) than the existing dataflow framework.

File: src/librustc_mir/dataflow/generic.rs (hunk 2 above)
Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/generic.rs]: Short answer: I hope to replace the current code that drives to the entry sets fixpoint with this code, but `BitDenotation` will remain (mostly) unchanged.

My long-term goal is to extend `Engine` with the graphviz debugging code, then replace the existing `DataflowBuilder` with it. Analyses that currently implement `BitDenotation` will continue to do so, and consumers of that data will have to wrap the analysis in the adapter described in [the summary](https://github.com/rust-lang/rust/pull/64566#issue-318580912) before passing it to `Engine` to be evaluated. In a perfect world, I would rename `BitDenotation` to `GenKillAnalysis`.

Before this happens, I'll also need to handle the `before_*` methods in the current `BitDenotation`. AFAICT, no analysis implements *both* the normal `effect` methods and their `before` variants, so we could do this by having every analysis implement the normal `effect` methods and have consumers decide whether those effects are observed at each statement by calling either `seek_before` or `seek_after`. However, while it is conceptually simpler, this seems more error-prone than having the analysis decide.

File: src/librustc_mir/dataflow/generic.rs (hunk 5 above)
Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/generic.rs]: It's not strictly necessary for most analyses since they also keep a reference to the `mir::Body`, but it hints to the user that the domain of the analysis depends on the body being analyzed.

One difference between `Analysis` and `BitDenotation` is that the various parts of the `mir::Body` (statements and terminators) are passed by reference to the `Analysis`. The old code required every analysis to store a reference to the body and do the lookup themselves. Simple analyses (like `HaveBeenBorrowedLocals`) wouldn't need to store a reference to the body as a field, they can operate on the same body that was passed to `Engine`. 

More complex analyses, like those that hold `MoveData` to deal with initialized places, have to do some computation ahead-of-time to figure out how big the domain is, and that computation will likely be shared between multiple analyses. They will ignore the `body` parameter here.

File: src/librustc_mir/dataflow/generic.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_mir/dataflow/generic.rs]: > In a perfect world, I would rename BitDenotation to GenKillAnalysis.

Seems entirely reasonable to do to me. Opinions @eddyb?

This explanation is great. Can you put it somewhere on the module docs, so ppl know how to decide which analysis scheme to use?

Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/generic.rs]: I wrote a summary of the current state in the latest commit with advice for people. Please take a look.

Comment (oli-obk) [reply src/librustc_mir/dataflow/generic.rs]: Thanks!

File: src/librustc_mir/dataflow/mod.rs (hunk 1 above)
Comment (Centril) [reply src/librustc_mir/dataflow/mod.rs]: @ecstatic-morse Since @oli-obk r+ed, maybe this can be done as a follow up?

Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/mod.rs]: Sounds good. I tentatively plan to do this in the `rustc` guide, then link to it from `dataflow/mod.rs`.

Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/mod.rs]: There's currently a "to be written" entry in the [appendix](https://rust-lang.github.io/rustc-guide/appendix/background.html?highlight=dataflow#what-is-a-dataflow-analysis)

**Comment by ecstatic-morse** [comment]
@oli-obk Can you r- and pick up the latest commit. There was a bug in the logic of `seek_after_assume_call_return`.

**Comment by ecstatic-morse** [comment]
@bors r-

(worth a shot)

File: src/librustc_mir/dataflow/mod.rs (hunk 1 above)
Comment (Centril) [reply src/librustc_mir/dataflow/mod.rs]: Sounds great. 👍
